### PR TITLE
made joiner.front assignable

### PR DESCRIPTION
**Pull request by Superstar64**

### CODE REVIEW

File: std/algorithm/iteration.d [hunk 1]
```diff
@@ -2415,6 +2415,15 @@ if (isInputRange!RoR && isInputRange!(ElementType!RoR))
                 return copy;
             }
         }
+
+        static if (hasAssignableElements!(ElementType!RoR))
+        {
+            @property void front(ElementType!(ElementType!RoR) element)
```
Comment (JakobOvrum) [inline std/algorithm/iteration.d]: `auto ref` perhaps?

Comment (Superstar64) [inline std/algorithm/iteration.d]: Don't auto ref args need to be templated? It didn't compile when I tried it.

Comment (JakobOvrum) [inline std/algorithm/iteration.d]: Right, they do.

Comment (JakobOvrum) [inline std/algorithm/iteration.d]: Maybe an overload in `hasLvalueElements` with a `ref` parameter would be appropriate?

**Comment by Superstar64** [comment]
I added a ref overload

**Comment by Superstar64** [comment]
I improved the unittest

**Comment by andralex** [comment]
I think just making `front` return `auto ref` should work, no?

**Comment by Superstar64** [comment]
This PR is for the case that the sub range has hasAssignableElements but does not have hasLvalueElements.

**Comment by JackStouffer** [comment]
Auto-merge toggled on

**Comment by JackStouffer** [comment]
@Superstar64 If reviewers go this long without commenting, please ping us!

Sometimes things fall through the cracks.

**Pull request merged**